PR: https://github.com/odoo/odoo/pull/160169

From: 02f756d061f1e52a2265e051ac57b13b311fb115
From: David Monnom (moda)
Date: 2024-04-29 23:21:31

Breaking data model changes scores: del:1 + add:0, change matches:
-    _inherit = 'pos.session'

Total Changes: 31

[IMP] pos_*: allow arbitrary data in related models

*: All pos related modules

This commit allows to access arbitrary data in related models. This is
useful for data that are computed on the server side and that are not
recomputed on the client side.

Before this commit, it was only possible to access data that are linked
to a field in the model.

Now custom data coming from the server can be added to any record with
the prefix `_` in the key. For example, if the server sends the
following data for `pos.order`:

Classic fields:
```
{
  id: 1,
  name: 'Order 1',
  partner_id: 1,
  partner_name: 'Partner 1',
}
```

Custom fields:
```
{
  id: 1,
  name: 'Order 1',
  partner_id: 1,
  partner_name: 'Partner 1',
  _custom_field: 'Custom value',
}
```

closes odoo/odoo#160169

Related: odoo/enterprise#60017
Signed-off-by: Adrien Guilliams (adgu) <adgu@odoo.com>

================================= pseudo patch: =================================

--- a/addons/pos_discount/models/__init__.py
+++ b/addons/pos_discount/models/__init__.py
@@ -3,3 +3,4 @@
 
 from . import pos_config
 from . import res_config_settings
+from . import product_product

--- a/addons/pos_discount/models/pos_session.py
+++ b/None
@@ -1,16 +0,0 @@
-# -*- coding: utf-8 -*-
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import models
-from odoo.osv.expression import OR
-
-
-class PosSession(models.Model):
-    _inherit = 'pos.session'
-
-    def _load_data_params(self, config_id):
-        params = super(PosSession, self)._load_data_params(config_id)
-        if self.config_id.module_pos_discount:
-            curr_domain = params['product.product']['domain']
-            params['product.product']['domain'] = OR([curr_domain, [('id', '=', self.config_id.discount_product_id.id)]])
-        return params

--- a/None
+++ b/addons/pos_discount/models/product_product.py
@@ -0,0 +1,14 @@
+from odoo import api, models
+from odoo.osv.expression import OR
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    @api.model
+    def _load_pos_data_domain(self, data):
+        params = super()._load_pos_data_domain(data)
+        config_id = self.env['pos.config'].browse(data['pos.config']['data'][0]['id'])
+        if config_id.module_pos_discount:
+            params = OR([params, [('id', '=', config_id.discount_product_id.id)]])
+        return params
